aurora/lrtools.py

```diff
@@ -0,0 +1,304 @@
+                                         p0, bounds=[(min_Te),(max_Te)],
+                                         loss='linear') # linear: standard least squares
+        
+    # robust way of obtaining uncertainties from output jacobian
+    U, s, Vh = np.linalg.svd(res_lsq.jac, full_matrices=False)
+    tol = np.finfo(float).eps*s[0]*max(res_lsq.jac.shape)
+    w = s > tol
+    cov = (Vh[w].T/s[w]**2) @ Vh[w]  # robust covariance matrix
+    perr = np.sqrt(np.diag(cov))     # 1 sigma uncertainty on fitted parameters
+
+    if fixed_ne_cm3:
+        ne_cm3 = fixed_ne_cm3
+        Te_eV = res_lsq['x'][0]
+        ne_cm3_unc = 0.
+        Te_eV_unc = perr[0]
+    else:
+        ne_cm3 = res_lsq['x'][0]*1e13
+        Te_eV = res_lsq['x'][1]
+        ne_cm3_unc = perr[0]*1e13
+        Te_eV_unc = perr[1]
+        
+    return ne_cm3, ne_cm3_unc, Te_eV, Te_eV_unc
+    
+        
+def plot_line_ratio_space(imp, cs, fun1_exc, fun2_exc, fun1_rec=None, fun2_rec=None,
+                          min_ne=1e13, max_ne=1e15, min_Te=0.1, max_Te=50):
+    '''Make a contour plot showing expected values of some line ratios over the (ne,Te) space.
+    Note that this only considers ionization equilibrium, i.e. no considerations on transport
+    are made.
+
+    Parameters
+    ----------
+    imp : str
+        Atomic symbol of chosen species.
+    cs : int
+        Integer representing the charge state of interest for excitation.
+        This is only used if both excitation and recombination components are considered.
+    fun1_exc : scipy.interpolate.RectBivariateSpline instance
+        Interpolation function for the excitation component of the first line, 
+        as given by :py:fun`~aurora.radiation.read_adf15`
+    fun2_exc : scipy.interpolate.RectBivariateSpline instance
+        Interpolation function for the excitation component of the second line, 
+        as given by :py:fun`~aurora.radiation.read_adf15`
+    fun1_rec : scipy.interpolate.RectBivariateSpline instance
+        Optional; recombination component of the first line.
+    fun2_rec : scipy.interpolate.RectBivariateSpline instance
+        Optional; recombination component of the second line.
+    min_ne, max_ne : floats
+        Bounds of ne grid [:math:`cm^{-3}`]
+    min_Te, max_Te : floats
+        Bounds of Te grid [:math:`eV`]
+
+    MWE:
+    ---
+    trs_NII = aurora.radiation.read_adf15('pec98#n_ssh_pju#n1.dat')
+    exc3995 = trs_NII[trs_NII['isel']==15]['log10 PEC fun'].iloc[0]
+    exc4042 = trs_NII[trs_NII['isel']==21]['log10 PEC fun'].iloc[0]
+    plot_line_ratio_space('N',1, exc4042, exc3995)
+    '''
+    # ne and Te grids  
+    ne_grid = np.linspace(min_ne, max_ne, 300)
+    Te_grid = np.linspace(min_Te, max_Te, 400)
+    NE, TE = np.meshgrid(ne_grid, Te_grid)
+
+    # fractional abundances at ionization equilibrium
+    atom_data = atomic.get_atom_data(imp, ["scd", "acd"])
+    _Te, fz = atomic.get_frac_abundances(
```

Comment:
Most of the code here should be possible to replace by two calls of class `tec `class